### ISSUE TITLE
Support for pointers to incomplete types 

### ISSUE THREAD

**Issue by mgieseki** [mpark/variant#34]
First of all, thank you for providing this implementation of `std::variant`. It's great to have  a lightweight alternative to `boost::variant` available in C++11.
I just stumbled over an issue with recursive type definitions, and I'm not sure whether it's a limitation of C++11 or the variant implementation. The following struct definition compiles fine with GCC's and Clang's C++17 support:

```cpp
struct S {
    std::variant<
        std::map<int, S>*
    > var;
};
```

When replacing `std::variant` with `mpark::variant`, I get an error message regarding the incomplete type `S` of `std::pair<int, S>::second`. Is it possible to enhance your implementation so that this kind of recursive type definition compiles with `mpark::variant` under C++11 as well?

**Comment by mpark** [comment]
Thanks for the report! I just tried it out and confirmed the issue. It either looks like my fallback implementation of `is_swappable` is instantiating too eagerly, or I'm checking for `is_swappable` where I shouldn't be. I'm going to be attending CppCon this week, so I'll be sure to fix this after the conference!

**Comment by mgieseki** [comment]
Awesome. Thanks for the confirmation and for having a look into the issue. Have a great time at CppCon.

**Comment by mpark** [comment]
The following commits address this issue:
  - https://github.com/mpark/variant/commit/e567a710ea6429b921fd42fa415d12232eab7e46
  - https://github.com/mpark/variant/commit/9e06a9d6a17fb31f2e4aabd42e59f48bd0ee0ed8
  - https://github.com/mpark/variant/commit/ea542a0fef86fb8c1696238b95bdaddb8ff2337c

**Comment by mgieseki** [comment]
Great! Thanks for fixing this issue. It works flawlessly now.

**Comment by cor3ntin** [comment]
I'm experiencing something similar with gcc 7, seems to work with 8 https://godbolt.org/z/0fb-bV
(also works if I use standard variant)
Thanks!

**Comment by mpark** [comment]
The core of the issue here was that my implementation of `invoke` was instantiating too much. This is fixed by 5713032c68c6dad70a6a1d9f72b80c737c249a28
